### PR TITLE
Update to version 1.2

### PR DESCRIPTION
**Pull request by Creeperface01**

### CODE REVIEW

**Comment by Creeperface01** [comment]
maybe @Snake1999 could merge all commits to increase commit count 😆

**Comment by LoYungSum** [comment]
Anyone have a direct download link?

**Comment by SupremeMortal** [comment]
https://ci.potestas.xyz/job/Nukkit-GT1.2/

**Comment by NycuRO** [comment]
@Creeperface01 
1) Potions not give effect and in Creative not work.
2) Banners aren't placeable.
3) Lingering potions not work.
4) Beacon not work.
5) Fireworks not work.
6) Armor Stands not work.
7) Totem of Undying not work.
8) Colored Beds when place for exemple yellow, it's blaced white.

image

IF YOU UNLOAD CHUNK AND LOAD AGAIN, THEM IT'S CHANGED:

image


9) Anvil durability not work, When you use them ,not changed to Slightly Broken or other. (On Creative test)
10) Potions aren't "placeable" in Cauldron. Cannot color water on Cauldron. Only put water.
11) You cannot enter on Minecart. (srsly, this broken 👎 )


**When you fix this, tell me, i'm come to Nukkit.**

**Comment by NycuRO** [comment]
[server.log](https://github.com/Nukkit/Nukkit/files/1362625/server.log)


Server.log if you want, with errors.

**Comment by Creeperface01** [comment]
@NycuRO these are not bugs, but unimplemented features, you can implement them as well as me

**Comment by NycuRO** [comment]
@Creeperface01 I'm say you because a part of them you "implemented" with commits on 1.2 branch.

**Comment by PikyCZ** [comment]
Bed Color bug is from 1.1update

File: src/main/java/cn/nukkit/form/element/ElementButton.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package cn.nukkit.form.element;
+
+public class ElementButton {
```
Comment (larryTheCoder) [inline src/main/java/cn/nukkit/form/element/ElementButton.java]: @Creeperface01 Is this normal... Or should it `extends Element`... or isn't?

Comment (Creeperface01) [reply src/main/java/cn/nukkit/form/element/ElementButton.java]: It shouldnt, because button can be added only to simple Windows, not custom

**Comment by Snake1999** [comment]
@NycuRO Do not expect everything on the same time :)

**Comment by NycuRO** [comment]
@Snake1999 I'm waiting one answer on Discord PR.

**Comment by nkeck720** [comment]
What's the progress like on this branch? I'm considering pulling this branch and using it as a 1.2 server to test it.

**Comment by Snake1999** [comment]
@boybook is working on GUI system, don't worry @nkeck720 :) But you can fork to test as it is stable enough for testing. 
For production use, we've got a big server to test it and fix bugs

**Comment by xISRAPILx** [comment]
Nice

**Comment by SupremeMortal** [comment]
@boybook there are two other fields after that. `motd second line` and `gamemode`

**Comment by Creeperface01** [comment]
hm MCPE crashes when you join the server now

**Comment by SupremeMortal** [comment]
Looking at other projects, instead of the `server uuid` in the RakNet UnconnectPong, they use the `hashcode` of the first motd line + `client address` + `client port`

**Comment by SupremeMortal** [comment]
@Creeperface01 Are all those Disc classes necessary? They have the same methods with different names. Seems like a lot of repeated code.

**Comment by Creeperface01** [comment]
Yes they are, If you dont want to change Item structure much.

**Comment by sero583** [comment]
@SupremeMortal @Creeperface01 I thought different disc's would have different meta's like wool blocks. But yes I agree with creeperface, the structure should stay how it is.

**Comment by TrevorVonSeggern** [comment]
Fyi I just tested the most recent build on my pi 3, and I couldn't connect from MCPE (v1.2.3).

File: src/main/java/cn/nukkit/network/protocol/PlaySoundPacket.java [hunk 2]
```diff
@@ -25,7 +25,7 @@ public void decode() {
     public void encode() {
         this.reset();
         this.putString(this.name);
-        this.putBlockVector3(this.x, this.y, this.z);
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/network/protocol/PlaySoundPacket.java]: Why 8?

File: src/main/java/cn/nukkit/network/protocol/PlaySoundPacket.java [hunk 3]
```diff
@@ -25,7 +25,7 @@ public void decode() {
...
+        this.putBlockVector3(this.x * 8, this.y * 8, this.z * 8);
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/network/protocol/PlaySoundPacket.java]: Why 8 there?

**Comment by Rsplwe** [comment]
emmmmmmmmm

**Comment by sero583** [comment]
@Creeperface01 agreed. Would recommened basically removing it out of inventory.

File: src/main/java/cn/nukkit/item/ItemArmor.java [hunk 4]
```diff
@@ -44,14 +44,19 @@ public boolean isArmor() {
     @Override
     public boolean onClickAir(Player player, Vector3 directionVector) {
         if (this.isHelmet() && player.getInventory().getHelmet().isNull()) {
-            if (player.getInventory().setHelmet(this)) this.setCount(0);
+            if (player.getInventory().setHelmet(this))
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/item/ItemArmor.java]: Why you don't use "{" here?

**Comment by Snake1999** [comment]
There're many servers who have a test server for 1.2, most of whom reported nukkit for 1.2.2 is somehow stable now. However, there are still some detailed client-side problem, like "unable to launch view mode", "no event after hot-bar item clicked" etc. 
After fixing it we will merge. Let's prepare our well welcome for 1.2.2 :)

**Comment by NycuRO** [comment]
Good 👍

**Comment by PeratX** [comment]
Oh. This has been a long way.

**Comment by SupremeMortal** [comment]
You should have just merged from upstream for the fix. Now there are conflicts.

File: src/main/java/cn/nukkit/item/ItemArmor.java [hunk 5]
```diff
@@ -38,6 +41,20 @@ public boolean isArmor() {
...
     }
 
+    @Override
+    public boolean onClickAir(Player player, Vector3 directionVector) {
+        if (this.isHelmet() && player.getInventory().getHelmet().isNull()) {
+            if (player.getInventory().setHelmet(this)) this.setCount(0);
+        } else if (this.isChestplate() && player.getInventory().getChestplate().isNull()) {
+            if (player.getInventory().setChestplate(this)) this.setCount(0);
+        } else if (this.isLeggings() && player.getInventory().getLeggings().isNull()) {
+            if (player.getInventory().setHelmet(this)) this.setCount(0);
```
Comment (Creeperface01) [inline src/main/java/cn/nukkit/item/ItemArmor.java]: #blameboybook for this

Comment (Goodstuff4UonYT) [reply src/main/java/cn/nukkit/item/ItemArmor.java]: xD

**Pull request merged**